### PR TITLE
Followup constraints tests

### PR DESCRIPTION
**Pull request by justus-camp**
## Description
This PR adds more test coverage for constraints.

## Reviewer Guidance
The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 1]
```diff
@@ -1012,6 +1038,199 @@ describe("Editing", () => {
...
+				const tree2 = tree.fork();
+
+				const rootPath: UpPath = {
+					parent: undefined,
+					parentField: rootFieldKeySymbol,
+					parentIndex: 0,
+				};
+
+				// Insert "a"
+				// State should be: [{ foo: "a"}]
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1100]:
```suggestion
				// tree: [{ foo: "a"}]
```

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 2]
```diff
@@ -1012,6 +1038,199 @@ describe("Editing", () => {
 
 				expectJsonTree([tree, tree2], [{ foo: "x" }]);
 			});
+
+			it("can put a constraint on newly inserted node", () => {
```
[the remark this answers is not in the document]
Comment (alex-pardes) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts]: I guess we're in the existence constraint describe block, so we probably don't need "existence constraint" in the name.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 3]
```diff
@@ -1012,6 +1038,199 @@ describe("Editing", () => {
...
+				rootSequence.insert(1, singleTextCursor({ type: jsonString.name, value: "b" }));
+				tree.transaction.commit();
+
+				// Concurrently delete a's ancestor
+				// State should be (to tree2): []
+				const tree2RootSequence = tree2.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				tree2RootSequence.delete(0, 1);
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1126]: This edit has to be sequenced before the transaction with the constraint if we want the constraint to be violated. When we just have two blocks of concurrent edits I find it less confusing to use tree/tree1 as the branch whose edits come first and tree2 as the branch whose edits come second.

Comment (alex-pardes) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1126]: But if you like you could keep the names and reverse the order of the merge/rebaseOnto calls below.

Comment (justus-camp) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1126]: Yeah I agree with your point about tree/tree1 having its edits first so I'll flip them. For some reason I thought this scenario should work with the order flipped and that we might need to make changes to support this.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 4]
```diff
@@ -1012,6 +1038,199 @@ describe("Editing", () => {
...
+
+				tree.merge(tree2);
+				tree2.rebaseOnto(tree);
+
+				expectJsonTree([tree, tree2], []);
+			});
+
+			it("constraint works with node created as part of transaction", () => {
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1134]:
```suggestion
			it("can add constraint to node inserted in same transaction", () => {
```

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 5]
```diff
@@ -1012,6 +1038,199 @@ describe("Editing", () => {
...
+				});
+				tree2RootSequence.insert(1, singleTextCursor({ type: jsonObject.name }));
+
+				tree.merge(tree2);
+				tree2.rebaseOnto(tree);
+
+				expectJsonTree([tree, tree2], [{ foo: {} }, {}, {}]);
+			});
+
+			it.skip("constraint on newly inserted node with ancestor deleted concurrently", () => {
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts]: Is this test meaningfully different from the other ancestor delete test? Was this supposed to be showing that a constraint that a node inserted in the same transaction exists can be violated if the insert fails due to an ancestor being deleted?

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 6]
```diff
@@ -1012,6 +1038,199 @@ describe("Editing", () => {
...
+				// State should be: [{ foo: "a" }]
+				const sequence = tree.editor.sequenceField({
+					parent: rootPath,
+					field: brand("foo"),
+				});
+				sequence.insert(0, singleTextCursor({ type: jsonString.name, value: "a" }));
+
+				// Constrain on "a" existing and insert "b" if it does
+				// State should be (if "a" exists): [{ foo: "a"}, "b"]
+				tree.transaction.start();
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1154]: I think the transaction is supposed to include the insert of "a".

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts (hunk 5 above)
Comment (justus-camp) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts]: That was the intent, but I think by swapping the order that makes the tests not meaningfully different if I'm understanding correctly.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 7]
```diff
@@ -1012,6 +1038,146 @@ describe("Editing", () => {
...
+
+				// Insert "a"
+				// State should be: ["a"]
+				const sequence = tree.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				sequence.insert(0, singleTextCursor({ type: jsonString.name, value: "a" }));
+
+				// Constrain on "a" and insert "b" into root sequence if it does
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1054]:
```suggestion
				// Insert "b" after "a" with constraint that "a" exists.
```

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 8]
```diff
@@ -1012,6 +1038,146 @@ describe("Editing", () => {
...
+					singleTextCursor({ type: jsonString.name, value: "c" }),
+				);
+
+				tree.merge(tree2);
+				tree2.rebaseOnto(tree);
+
+				expectJsonTree([tree, tree2], ["c", "a", "b"]);
+			});
+
+			it("can add constraint to node inserted in same transaction", () => {
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1086]: We should have a variant of this test demonstrating that the transaction fails if the node with the constraint cannot be inserted.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 9]
```diff
@@ -1012,6 +1038,146 @@ describe("Editing", () => {
...
+						parent: secondPath,
+					},
+					0,
+				);
+
+				const rootSequence = tree.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				rootSequence.delete(0, 1);
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1161]: I think we wanted to test rebasing over a commit which simultaneously moves the target node and deletes its original parent. To test that, these two edits would need to be in the same transaction.

Comment (alex-pardes) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1161]: We should probably also have a variant of this test which verifies that simultaneously moving the target node and deleting its new parent causes a constraint violation.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 10]
```diff
@@ -1012,6 +1038,199 @@ describe("Editing", () => {
 
 				expectJsonTree([tree, tree2], [{ foo: "x" }]);
 			});
+
+			it("can put a constraint on newly inserted node", () => {
+				const tree = makeTreeFromJson([{}]);
```
[the remark this answers is not in the document]
Comment (justus-camp) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts]: Updated this test to use a flat sequence.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts (hunk 6 above)
Comment (justus-camp) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts]: Moved "a" insert to be inside the transaction.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts (hunk 8 above)
Comment (justus-camp) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1086]: Added test for this scenario that is currently failing and skipped. Will follow with a PR that fixes it.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts (hunk 9 above)
Comment (justus-camp) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1209]: Moved the edit to be in the same transaction and also wrote another cross-field move test as described. It's skipped for now for a similar issue as the other one and will be fixed in a follow-up PR.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 11]
```diff
@@ -1012,6 +1038,257 @@ describe("Editing", () => {
...
+				rootSequence.insert(1, singleTextCursor({ type: jsonString.name, value: "b" }));
+				tree2.transaction.commit();
+
+				tree.merge(tree2);
+				tree2.rebaseOnto(tree);
+
+				expectJsonTree([tree, tree2], []);
+			});
+
+			it("cross-field move test", () => {
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1178]: A more descriptive test name would be nice. Maybe something like "not violated by move out under delete".

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 12]
```diff
@@ -1012,6 +1038,257 @@ describe("Editing", () => {
...
+
+				tree.merge(tree2);
+				tree2.rebaseOnto(tree);
+
+				expectJsonTree([tree, tree2], [{ foo2: ["a"] }, {}]);
+			});
+
+			// TODO: Constraint state isn't updated properly because
+			// rebaseChild isn't called when currMark is undefined in rebaseMarkList
+			it.skip("cross-field move delete violates constraint", () => {
```
Comment (alex-pardes) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1233]: This one could be called something like "violated by move in under delete".

**Review by alex-pardes — APPROVED** [review]

**Pull request merged**